src/hotspot/share/opto/mulnode.cpp

```diff
@@ -970,6 +970,40 @@ static int maskShiftAmount(PhaseGVN* phase, Node* shiftNode, int nBits) {
   return 0;
 }
 
+// (X << con1) << con0 with con0 < nbits && con1 < nbits ==>
+// if con0 + con1 >= nbits => 0
+// if con0 + con1 < nbits => X << (con1 + con0)
+//
+// outer_shift is, with the notation above, the (...) << con0
+// con0 is the rhs of outer_shift (since it's already computed in the callers)
+// con0 is assumed to be masked already (as computed by maskShiftAmount) and non-zero
+// bt must be T_LONG or T_INT.
+static Node* collapse_nested_shift_left(PhaseGVN* phase, Node* outer_shift, int con0, BasicType bt) {
+  assert(bt == T_LONG || bt == T_INT, "Unexpected type");
+  int nbits = bt == T_LONG ? BitsPerJavaLong : BitsPerJavaInteger;
```

Comment:
Roland is introducing a new method for this in `https://github.com/openjdk/jdk/pull/23438`, see `bits_per_java_integer`. I suggest you use it too ;)

Reply:
Happily, as soon as this other PR is merged!

Reply:
Or you just copy it and hope that the merge will eventually be clean, up to you.